### PR TITLE
V3 tuples

### LINKED ISSUE #177: Tuple type (2.1)
Final new 2.1 feature that we're looking at (https://issues.apache.org/jira/browse/CASSANDRA-6875) requires the new CQL tuple type: https://issues.apache.org/jira/browse/CASSANDRA-7248

### PR DESCRIPTION
**Pull request by Zariel**
closes #177

### CODE REVIEW

**Comment by Zariel** [comment]
This isnt really as large as it looks like, the problem is that `TypeInfo` is scattered around so many places, which is unfortunate. As noted in the commit this will break implementations of Un/MarshallCQL

File: marshal.go [hunk 1]
```diff
@@ -25,22 +25,22 @@ var (
 // Marshaler is the interface implemented by objects that can marshal
 // themselves into values understood by Cassandra.
 type Marshaler interface {
-	MarshalCQL(info *TypeInfo) ([]byte, error)
+	MarshalCQL(info TypeInfo) ([]byte, error)
```
Comment (0x6e6562) [inline marshal.go]: Why is `TypeInfo` no longer a pointer?

File: marshal.go [hunk 2]
```diff
@@ -1148,29 +1169,107 @@ func unmarshalInet(info *TypeInfo, data []byte, value interface{}) error {
...
-	Elem   *TypeInfo // only used for TypeMap, TypeList and TypeSet
-	Custom string    // only used for TypeCustom
+type TypeInfo interface {
+	Type() Type
+	Version() byte
+	Custom() string
+
+	// New creates a pointer to an empty version of whatever type
+	// is referenced by the TypeInfo receiver
+	New() interface{}
```
Comment (0x6e6562) [inline marshal.go]: Is `New()` as function prototype idiomatic Go?

File: session.go [hunk 3]
```diff
@@ -632,20 +632,46 @@ func (iter *Iter) Scan(dest ...interface{}) bool {
...
+			i++
 			continue
 		}
-		err := Unmarshal(iter.columns[i].TypeInfo, iter.rows[iter.pos][i], dest[i])
-		if err != nil {
-			iter.err = err
+
+		// how can we allow users to pass in a single struct to unmarshal into
+		if col.TypeInfo.Type() == TypeTuple {
+			// this will panic, actually a bug, please report
```
Comment (0x6e6562) [inline session.go]: What does the comment about the panic mean?

File: session.go [hunk 4]
```diff
@@ -632,20 +632,46 @@ func (iter *Iter) Scan(dest ...interface{}) bool {
...
+			iter.err = Unmarshal(col.TypeInfo, iter.rows[iter.pos][c], dest[i:i+count])
+			if iter.err != nil {
+				return false
+			}
+			i += count
+			continue
+		}
+
+		iter.err = Unmarshal(col.TypeInfo, iter.rows[iter.pos][c], dest[i])
+		if iter.err != nil {
```
Comment (0x6e6562) [inline session.go]: I imagine this return value would be different in #331 had landed already?

**Comment by 0x6e6562** [comment]
On the whole LGTM - the patch is noisy because you've plumbed through a change to the central type hierarchy of the library, but those changes are mainly mechanical.

I've just remarked on a few items of a minor nature.

File: marshal.go (hunk 1 above)
Comment (Zariel) [inline marshal.go]: It was changed to an interface

File: marshal.go (hunk 2 above)
Comment (Zariel) [inline marshal.go]: No, but its the best way I could think of the keep the current behaviour of the helper functions, which I didnt want to change in this patch.

File: session.go (hunk 3 above)
Comment (Zariel) [inline session.go]: If we're unmarshalling a tuple but the `TypeInfo` isn't a tuple type then the type assertion will fail.

File: marshal.go (hunk 1 above)
Comment (0x6e6562) [inline marshal.go]: Ah yes, you're right :-)

File: session.go (hunk 3 above)
Comment (0x6e6562) [inline session.go]: I don't get that - which invocation are you talking about:

```go
if col.TypeInfo.Type() == TypeTuple {
```

or

```go
tuple := col.TypeInfo.(TupleTypeInfo)
```

?

File: marshal.go (hunk 2 above)
Comment (0x6e6562) [inline marshal.go]: OK, fair enough.

File: session.go (hunk 3 above)
Comment (Zariel) [inline session.go]: `tuple := col.TypeInfo.(TupleTypeInfo)`

Comment (0x6e6562) [inline session.go]: If you think that is possible, why not be defensive, check the type cast and return some error in the failure scenario:

```go
tuple, ok := col.TypeInfo.(TupleTypeInfo)
if !ok {
    return an error ....
}
```

Comment (Zariel) [inline session.go]: Then it breaks the `iter.Scan()` interface which would not be ideal, this case should not happen in practice, I left that comment there to make it explicit that it wasn't doing the safe type assertion.

File: session.go (hunk 4 above)
Comment (Zariel) [inline session.go]: I don't think so, as it should already be set before the call the `Scan()`

**Comment by 0x6e6562** [comment]
Nice work :-)

**Comment by Zariel** [comment]
v3 support is coming along nicely :+1:

**Pull request merged**